hook-worker/src/worker.rs

```diff
@@ -162,68 +177,94 @@ impl<'p> WebhookWorker<'p> {
                 .set(1f64 - semaphore.available_permits() as f64 / self.max_concurrent_jobs as f64);
         };
 
+        let dequeue_batch_size_histogram = metrics::histogram!("webhook_dequeue_batch_size");
+
         if transactional {
             loop {
                 report_semaphore_utilization();
-                let webhook_job = self.wait_for_job_tx().await;
-                spawn_webhook_job_processing_task(
-                    self.client.clone(),
-                    semaphore.clone(),
-                    self.retry_policy.clone(),
-                    webhook_job,
-                )
-                .await;
+                let mut batch = self.wait_for_jobs_tx().await;
+                dequeue_batch_size_histogram.record(batch.jobs.len() as f64);
+
+                // Get enough permits for the jobs before spawning a task.
```

Comment:
We might be blocked here due to one missing permit.

We might get more throughput if we only pull `min(semaphore.available_permits(), dequeue_batch_size)`, so that we guarantee we won't block here.

Reply:
Don't we want to block?

In my first stab at this code, I pushed the permit grabbing down into `process_webhook_job`, so each job would grab its own permit before making the HTTP request. The problem is that puts no backpressure on how many tasks we spawn.

If we use `min(semaphore.available_permits(), dequeue_batch_size)` then we could spawn off the work of 100 jobs by only grabbing 1 permit. If we do that then what's the point of having `max_concurrent_jobs` at all?

Maybe `max_concurrent_jobs` can be removed in-lieu of leaning on the connection count configurable. I think the permit system matters more for non-transactional mode.

Reply:
Placed the comment here because this is where we block, but my recommendation applies higher: change the signature of `wait_for_jobs()` to accept a max job count to pull, so that we can call `wait_for_jobs(min(semaphore.available_permits(), dequeue_batch_size))`.

That's something we can implement later on, we can skip it for now and revisit if we manage to actually hit this semaphore.

Reply:
Ah-ha, gotcha, that makes sense.

Reply:
Again, I think it's fine to just add a `TODO` comment when calling `wait_for_jobs()`, and move on for now.